rustworkx-core/src/dag_algo.rs

```diff
@@ -314,6 +316,140 @@ where
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected // Used for .edged_directed
+        + GraphBase<NodeId = N>,
+    N: Debug + Copy + Eq + Hash,
+{
+    type Item = Vec<N>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.first_iter {
+            self.first_iter = false;
+            self.cycle_check.insert(self.cur_layer.clone());
+            Some(self.cur_layer.to_owned())
+        } else if self.cur_layer.is_empty() {
+            None
+        } else {
+            for node in &self.cur_layer {
+                if self.graph.to_index(*node) >= self.graph.node_bound() {
+                    panic!("Node {:#?} is not present in the graph.", node);
+                }
+                let children = self
+                    .graph
+                    .neighbors_directed(*node, petgraph::Direction::Outgoing);
+                let mut used_indices: HashSet<G::NodeId> = HashSet::new();
+                for succ in children {
+                    // Skip duplicate successors
+                    if used_indices.contains(&succ) {
+                        continue;
+                    }
+                    used_indices.insert(succ);
+                    let mut multiplicity: usize = 0;
+                    let raw_edges: G::EdgesDirected = self
+                        .graph
+                        .edges_directed(*node, petgraph::Direction::Outgoing);
+                    for edge in raw_edges {
+                        if edge.target() == succ {
+                            multiplicity += 1;
+                        }
+                    }
+                    self.predecessor_count
+                        .entry(succ)
+                        .and_modify(|e| *e -= multiplicity)
+                        .or_insert(
+                            // Get the number of incoming edges to the successor
+                            self.graph
+                                .edges_directed(succ, petgraph::Direction::Incoming)
+                                .count()
+                                - multiplicity,
+                        );
+                    if *self.predecessor_count.get(&succ).unwrap() == 0 {
+                        self.next_layer.push(succ);
+                        self.predecessor_count.remove(&succ);
+                    }
+                }
+            }
+            swap(&mut self.cur_layer, &mut self.next_layer);
+            if self.cycle_check.contains(&self.cur_layer) {
+                panic!("The layers algorithm doesn't support graphs with cycles.")
+            }
+            self.cycle_check.insert(self.cur_layer.to_owned());
+            self.next_layer.clear();
+            if self.cur_layer.is_empty() {
+                None
+            } else {
+                Some(self.cur_layer.to_owned())
+            }
+        }
+    }
```

Comment:
I still have some concerns over the cycle checking here. This is much better than calling `find_cycle` but this potentially involves 2 copies which feels like one too many. I know we talked about this offline but I still feel like the cycle checking should be by node, not by layer. If you're getting duplicates here that's pointing to a bug somewhere, likely in the user provided first layer.

Reply:
Yes I will be updating it to be by nodes instead, I just haven't modified the code to make it work that way. I haven't played with this since we spoke offline 😅. I am actively thinking of which ways to modify this method to make it work as intented. I will keep you posted on this.